docs/custom-plugins/custom-decorators.md

```diff
@@ -0,0 +1,158 @@
+# Decorators in plugins
+
+Decorators transform API descriptions, by adding, removing, or changing elements of the document. Before you build your own decorators:
+
+* [Learn about Redocly decorators](../decorators.md).
+* [Check the list of built-in decorators](../decorators.md#list-of-decorators).
+
+If you can't find an existing decorator that fits your needs, then you can add a decorator in a plugin.
+
+:::warning Preprocessors
+
+Decorators and preprocessors are the same in structure, but preprocessors are run _before_ linting, and decorators are run after. We always recommend using decorators where possible, since the document might not be valid or structured as expected if the linting step hasn't run yet.
+
+:::
+
+## Plugin structure
+
+To create a preprocessor or decorator, the object that is exported from your module has to conform to an interface such as the following example:
+
+```js
+module.exports = {
+  id: 'my-local-plugin',
+  preprocessors: {
+    oas3: {
+      "processor-id": () => {
+        // ...
+      }
+    }
+  },
+  decorators: {
+    oas3: {
+      "decorator-id": () => {
+        // ...
+      }
+    }
+  }
+}
+```
+
+Each decorator or preprocessor is a function that returns an object. The object's keys are the node types in the document, and each of those can contain any or all of the `enter()`, `leave()` and `skip()` functions for that node type. Find more information and examples on the [visitor pattern page](./visitor.md).
+
+## Decorator example
+
+To give a small (but fun) example, here is a decorator that adds a sparkle emoji ✨ at the start of every operation description.
+
+To help keep the plugin code organized, this example uses one file per decorator. In this example, this is the file `plugins/decorators/operation-sparkle.js`:
+
+```js
+module.exports = OperationSparkle;
+
+function OperationSparkle() {
+  console.log("adding sparkles ... ");
+  return {
+    Operation: {
+      leave(target) {
+        if(target.description) {
+          target.description = "✨ " + String(target.description);
+        }
+      }
+    },
+  }
+};
+```
+
+Decorators use the [visitor pattern](./visitor.md) to run an operation on every node in the document. In this example, when the code executes the `leave()` function on the `Operation` node, it checks if the node (passed as `target` in this example) has a description, and updates it if it does.
```

Comment:
question: Do you have to use `target` or can it be replaced with any string, such as `op`? I'm not clear for example if this also works:

```js
module.exports = OperationSparkle;

function OperationSparkle() {
  console.log("adding sparkles ... ");
  return {
    Operation: {
      leave(op) {
        if(op.description) {
          op.description = "✨ " + String(op.description);
        }
      }
    },
  }
};
```

Reply:
yes, I tried to use "target" in an attempt to make it a bit more copy/paste friendly, but called out that it was just an example, so I think it's clear.